feat(actions): Pluggable DQ Actions & Alerting (#1289)
databrickslabs/dqx: docs/dqx/docs/guide/actions_and_alerts.mdx

---
sidebar_position: 311
---

import Admonition from '@theme/Admonition';
import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

# Actions and Alerting

> DQX **actions** let you react automatically when observed summary metrics cross a threshold. After checks are applied and the Spark observation is materialized, each action evaluates its optional *condition* against the observed metrics and, when the condition is satisfied (or when there is no condition), executes its inner logic.

Reviewer: Might be helpful to add a small introductory note about when to use actions. Something like:

```suggestion
DQX **actions** let you react automatically when observed summary metrics cross a threshold. Actions can be used to notify stakeholders or fail a pipeline when data quality issues are detected.

After checks are applied and the Spark observation is materialized, each action evaluates its optional *condition* against the observed metrics and, when the condition is satisfied (or when there is no condition), executes its inner logic.
```

The built-in action types are:

> - *DQAlert* — delivers a notification to one or more destinations (Slack, Microsoft Teams, generic HTTPS webhook, or an in-process callback).

Reviewer: Should we mention logging here as well?

Author: added

from databricks.labs.dqx.engine import DQEngine
from databricks.labs.dqx.metrics_observer import DQMetricsObserver
from databricks.labs.dqx.config import InputConfig, OutputConfig
from databricks.labs.dqx.actions import DQActionManager

observer = DQMetricsObserver(name="my_pipeline")

# Load the actions defined in actions.yml
actions = DQActionManager.load_actions_from_local_file("actions.yml")

# Actions require an observer
engine = DQEngine(WorkspaceClient(), observer=observer, actions=actions)

# Actions fire automatically after the save completes
engine.apply_checks_and_save_in_table(
    checks=checks,
    input_config=InputConfig(location="catalog.schema.input"),
    output_config=OutputConfig(location="catalog.schema.valid"),
    quarantine_config=OutputConfig(location="catalog.schema.quarantine"),
    metrics_config=OutputConfig(location="catalog.schema.metrics"),
)
```

  </TabItem>
</Tabs>

<Admonition type="warning" title="Observer required">
Passing *actions* to *DQEngine* without also passing an *observer* raises `InvalidParameterError` at construction time. Actions evaluate metrics produced by the observer and cannot function without one.

> </Admonition>

Reviewer: I think this section is combining 2 concepts that might be clearer if split:

1. Defining an action when initializing `DQEngine`
2. Checking data with end-to-end methods to generate an action

I would suggest adding a small section like `### Configuring actions` before this one:

```
### Configuring actions

To run actions, attach them to a `DQEngine` together with a `DQMetricsObserver`. 
`DQEngine` accepts `DQAction` instances, raw metadata dicts, or a list loaded from 
a file with `DQActionManager` — and mixed lists of instances and dicts are supported.

... CODE BLOCK ...
```

Author: added

retries) therefore blocks that *callback* rather than data processing directly; the practical risk
is that a slow callback lets Spark's listener event queue back up and **drop progress events**,
which would skip action evaluation for those batches. Keep the webhook *timeout* and retry budget
small so this cannot happen. A non-terminal delivery failure is logged and never crashes the stream.

`FailPipeline`'s immediate-abort guarantee applies to the **batch** path (where actions are evaluated
synchronously in the save call). In **streaming**, `FailPipeline` is raised inside the listener
callback, and Spark generally isolates and logs listener exceptions rather than stopping the query —
so do not rely on `FailPipeline` to hard-stop a streaming query; use a downstream gate if you need a
guaranteed stop.

In practice, alerts are rate-limited (`HOURLY` / `DAILY`) and typically fire only when a condition
crosses a threshold, so the callback rarely blocks and the synchronous path is sufficient. Fully
asynchronous off-thread delivery is a possible future enhancement, mainly for high-throughput
streaming that fires alerts frequently against slow endpoints.

> </Admonition>

Reviewer: Because the condition is optional, I think we should suggest that users specify a condition or frequency for continuous streaming data.

Author: added

### Manual evaluation with evaluate_actions

For pipelines that do not use the save methods (e.g. Lakeflow Pipelines or custom `writeStream` code), call `evaluate_actions` explicitly after triggering a Spark action on the checked DataFrame:

<Tabs>
  <TabItem value="Python" label="Python (classes)" default>

```python
from databricks.sdk import WorkspaceClient
from databricks.labs.dqx.engine import DQEngine
from databricks.labs.dqx.metrics_observer import DQMetricsObserver
from databricks.labs.dqx.actions import DQAction, DQAlert, SlackDQAlertDestination

observer = DQMetricsObserver(name="manual_eval")
engine = DQEngine(WorkspaceClient(), observer=observer, actions=[
    DQAction(
        condition="error_row_count > 0",
        action=DQAlert(
            destinations=[SlackDQAlertDestination(name="slack", webhook_url="https://hooks.slack.com/…")]
        ),
    )
])

# Apply checks — returns the DataFrame and an observation handle
checked_df, observation = engine.apply_checks_by_metadata(input_df, checks)

# Trigger a Spark action to materialize the observation

> checked_df.count()  # or write to a table, etc.

Reviewer: It might be useful to show practical examples of which Spark actions to trigger for specific behavior. 

These use cases come to mind:

1. For notifications prefer `write`, then `evaluate_actions` to avoid an unnecessary `count`.
2. To fail the pipeline, use `count`, then `evaluate_actions` to fail before writing bad data.

Author: added

</Tabs>

Compound conditions using `and` / `or` are supported:

<Tabs>
  <TabItem value="Python" label="Python (classes)" default>

```python
DQAction(
    condition="error_row_count > 0 or warning_row_count > 0",
    action=DQAlert(destinations=[SlackDQAlertDestination(name="slack", webhook_url="https://hooks.slack.com/…")]),
)
```

  </TabItem>
  <TabItem value="YAML" label="YAML (metadata)">

```yaml
- action:
    type: alert
    destinations:
      - type: slack
        name: slack
        webhook_url: https://hooks.slack.com/…
  condition: error_row_count > 0 or warning_row_count > 0
```

  </TabItem>
</Tabs>

## Action Types

Every action is wrapped in a `DQAction` that binds it to an optional [condition](#condition-syntax-and-available-metrics); the action runs when the condition is met (or when no condition is set). DQX ships three built-in action types, each documented in its own section below:

- [DQAlert](#dqalert) — send a notification to one or more destinations.
- [FailPipeline](#failpipeline) — abort the current run by raising `PipelineFailedError`.
- [NoOpAction](#noopaction-dry-run--audit) — fire without any side effect (testing / dry-run / audit).

### DQAlert

> `DQAlert` delivers a notification to one or more destinations (Slack, Microsoft Teams, generic HTTPS webhook, log, or an in-process callback). A single alert message is dispatched concurrently to every configured destination, and a delivery failure in one destination is isolated — the others still receive the alert, and per-destination errors are captured in `ActionResult.destination_errors`. The available destinations are listed under [Alert types](#alert-types), and [Controlling when alerts are sent](#controlling-when-alerts-are-sent) covers rate-limiting and status-change-only delivery.

Reviewer: Maybe we should make it explicit that these are the Spark driver logs.

Author: fixed

<Tabs>
  <TabItem value="Python" label="Python (classes)" default>

```python
from databricks.labs.dqx.actions import DQAction, DQAlert, SlackDQAlertDestination

alert = DQAction(
    condition="error_row_count > 0",
    action=DQAlert(destinations=[SlackDQAlertDestination(name="slack", webhook_url="https://hooks.slack.com/…")]),
)
```

  </TabItem>
  <TabItem value="YAML" label="YAML (metadata)">

```yaml
- action:
    type: alert
    destinations:
      - type: slack
        name: slack
        webhook_url: https://hooks.slack.com/…
  condition: error_row_count > 0
```

  </TabItem>
</Tabs>

> ### FailPipeline

Reviewer: Do we need a note or warning about how this will behave for streaming? The listener and query evaluation are independent, so we can't guarantee any immediate stopping.

Author: added

<Tabs>
  <TabItem value="Python" label="Python (classes)" default>

```python
from databricks.labs.dqx.actions import DQAlert, WebhookDQAlertDestination
from databricks.labs.dqx.config import DQSecret

alert = DQAlert(
    destinations=[
        WebhookDQAlertDestination(
            name="pagerduty",
            webhook_url="https://events.pagerduty.com/v2/enqueue",
            # In production, use DQSecret — see the Secrets section below
            username=DQSecret(scope="dq-secrets", key="pd-user"),
            password=DQSecret(scope="dq-secrets", key="pd-token"),
        )
    ]
)
```

  </TabItem>
  <TabItem value="YAML" label="YAML (metadata)">

```yaml
- action:
    type: alert
    destinations:
      - type: webhook
        name: pagerduty
        webhook_url: https://events.pagerduty.com/v2/enqueue
        # In production, reference secrets rather than plaintext credentials
        username:
          secret: dq-secrets/pd-user
        password:
          secret: dq-secrets/pd-token
```

  </TabItem>
</Tabs>

### Log (driver logger)

> `LogDQAlertDestination` writes the alert to the standard Python logger on the Spark driver instead of contacting an external system. Because it performs no network I/O it never triggers SSRF validation and cannot stall a streaming micro-batch, which makes it the ideal destination for local development, demos, and end-to-end tests. Unlike the callback destination it is fully serializable, so it has a `type: log` metadata form and can be persisted. The optional `level` controls the log level (`debug`, `info`, `warning`, `error`, or `critical`; defaults to `warning`).

Reviewer: I think the part about serializability might be unnecessary here?

```suggestion
`LogDQAlertDestination` writes the alert to the standard Python logger on the Spark driver instead of contacting an external system. Because logging performs no network I/O, it never triggers SSRF validation and cannot stall a streaming micro-batch. This makes it an ideal destination for local development, demos, and end-to-end tests. The optional `level` controls the log level (`debug`, `info`, `warning`, `error`, or `critical`; defaults to `warning`).
```

Author: true
